dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamAggregationTransforms.java

```diff
@@ -134,545 +134,201 @@ public Instant apply(BeamSqlRow input) {
-            break;
-          case TINYINT:
-            finalRecord.addField(idx, (byte) (accumulator.getByte(idx) / count));
-            break;
-          case FLOAT:
-            finalRecord.addField(idx, (float) (accumulator.getFloat(idx) / count));
-            break;
-          case DOUBLE:
-            finalRecord.addField(idx, accumulator.getDouble(idx) / count);
-            break;
-          default:
-            break;
-          }
-          break;
-        case "SUM":
-          switch (ex.getOutputType()) {
-          case INTEGER:
-            finalRecord.addField(idx, accumulator.getInteger(idx));
-            break;
-          case BIGINT:
-            finalRecord.addField(idx, accumulator.getLong(idx));
-            break;
-          case SMALLINT:
-            finalRecord.addField(idx, accumulator.getShort(idx));
-            break;
-          case TINYINT:
-            finalRecord.addField(idx, accumulator.getByte(idx));
-            break;
-          case FLOAT:
-            finalRecord.addField(idx, accumulator.getFloat(idx));
-            break;
-          case DOUBLE:
-            finalRecord.addField(idx, accumulator.getDouble(idx));
-            break;
-          default:
-            break;
-          }
-          break;
-        case "MAX":
-        case "MIN":
-          switch (ex.getOutputType()) {
-          case INTEGER:
-            finalRecord.addField(idx, accumulator.getInteger(idx));
-            break;
-          case BIGINT:
-            finalRecord.addField(idx, accumulator.getLong(idx));
-            break;
-          case SMALLINT:
-            finalRecord.addField(idx, accumulator.getShort(idx));
-            break;
-          case TINYINT:
-            finalRecord.addField(idx, accumulator.getByte(idx));
-            break;
-          case FLOAT:
-            finalRecord.addField(idx, accumulator.getFloat(idx));
-            break;
-          case DOUBLE:
-            finalRecord.addField(idx, accumulator.getDouble(idx));
-            break;
-          case TIMESTAMP:
-            finalRecord.addField(idx, accumulator.getDate(idx));
-            break;
-          default:
-            break;
-          }
-          break;
-        default:
-          break;
+    public List<BeamSqlRow> addInput(List<BeamSqlRow> accumulator, BeamSqlRow input) {
+      List<BeamSqlRow> deltaAcc = new ArrayList<>();
+      for (int idx = 0; idx < aggregators.size(); ++idx) {
+        deltaAcc.add(aggregators.get(idx).add(accumulator.get(idx),
+            sourceFieldExps.get(idx).evaluate(input).getValue()));
+      }
+      return deltaAcc;
+    }
+    @Override
+    public List<BeamSqlRow> mergeAccumulators(Iterable<List<BeamSqlRow>> accumulators) {
+      List<BeamSqlRow> deltaAcc = new ArrayList<>();
+      for (int idx = 0; idx < aggregators.size(); ++idx) {
+        List<BeamSqlRow> accs = new ArrayList<>();
+        while (accumulators.iterator().hasNext()) {
```

Comment:
Use a for loop.

Also note that Iterable.iterator() always returns a new iterator, so you need to store it in a variable before calling hasNext() and next() unless you're only trying to get the very first element. There are a bunch of instances of this across this PR that should be fixed.